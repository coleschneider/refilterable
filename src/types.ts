import { History } from 'history';
import { MutableRefObject } from 'react';

export const CREATE_FILTER_MARKER = Symbol("createFilter");

export interface FilterConfig<T = string> {
	parse?(input: string): T; 
  format?(value: T): string; 
  validate?(input: string, parse: (input: string) => T): boolean;
  defaultValue?: T | undefined;
}

export interface FilterObject<T = string> extends FilterConfig<T> {
  paramName: string;
  parse(input: string): T; 
  format(value: T): string; 
  validate(input: string, parse: (input: string) => T): boolean;
}

/**
 * User defined type guard to check if the passed object is a valid instance of FiterObject
 * 
 * @param filter 
 */
export function isFilterObject<T>(filter: any): filter is FilterObject<T> {
  return filter[CREATE_FILTER_MARKER];
}

export type FiltersContextValue = {
  locationObserver: MutableRefObject<{ 
    watch(paramName: string, callback: Function): Function;
    getParamInfo(paramName: string): { hasParam: boolean, paramValue: any };
    getCurrentParams(): URLSearchParams;
  }>;
  history?: History;
  filterRegistry: Map<string, FilterObject<any>>,
}

export type ForwardHistoryAction = 'PUSH' | 'REPLACE';

export type SetFilterOptions = {
  dry?: boolean;
  action?: ForwardHistoryAction;
};

export type FilterSetter<T = string> = (nextValue: T, options?: SetFilterOptions) => string;
