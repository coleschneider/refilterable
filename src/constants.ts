import { SetFilterOptions, ForwardHistoryAction } from './types';

export const defaultSetFilterOptions: SetFilterOptions = Object.freeze({
  dry: false,
  action: 'PUSH' as ForwardHistoryAction,
});