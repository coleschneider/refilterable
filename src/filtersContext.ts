import { createContext } from "react";
import { FiltersContextValue } from './types';
import createLocationObserver from "./createLocationObserver";

export default createContext<FiltersContextValue | null>(null);