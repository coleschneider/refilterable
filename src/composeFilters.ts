import { FILTER_COMPOSITION_MARKER } from './utils/constants';
import { FilterObject, FilterComposition } from './utils/types';

// @todo: support filter compositions in filters
export default function composeFilters(filters: FilterObject[], validate?: (input: any) => boolean): FilterComposition {
  return ({
    filters,
    validate: typeof validate === 'function' ? validate : () => true,
    [FILTER_COMPOSITION_MARKER]: true,
  } as FilterComposition) ;
}