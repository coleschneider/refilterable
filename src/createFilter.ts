import invariant from 'invariant';
import { FilterConfig, FilterObject } from './utils/types';
import { CREATE_FILTER_MARKER } from './utils/constants';


export default function createFilter<T>(paramName: string, config?: FilterConfig<T>): FilterObject<T> {
  invariant(
    String(paramName || '').length > 0,
    `re-filter: param name cannot be empty`,
  );

  const defaultConfig = {
    parse(input: string): any { return input; }, 
    format(value: T): string { return String(value); },
    validate(): boolean { return true; },
    defaultValue: undefined,
    [CREATE_FILTER_MARKER]: true,
  };

  return {
    ...defaultConfig,
    ...config,
    paramName,
  };
}; 