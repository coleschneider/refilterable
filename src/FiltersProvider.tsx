import React, { useEffect, useRef, FunctionComponent } from 'react';
import invariant from 'invariant';
import { History, Location } from 'history';
import filtersContext from './filtersContext';
import { FiltersContextValue, FilterObject } from './types';
import createLocationObserver from './createLocationObserver';
 
type Props = {
  history: History;
  children: JSX.Element[];
}

const FiltersProvider: FunctionComponent<Props> = ({ history, children }: Props) => {
  invariant(
    history,
    `re-filter: FiltersProvider was not passed a history instance. 
     Make sure you pass it via the history prop`,
  );

  const filterRegistry = useRef<Map<string, FilterObject<any>>>(new Map());
  const locationObserver = useRef(createLocationObserver(history.location.search));

  useEffect(() => {
    // listen to location changes
    const unlisten = history.listen((location: Location) => {
      locationObserver.current.notify(location.search);
    });

    // unsubscribe when the component unmounts
    return unlisten;
  }, []);

  const contextValue: FiltersContextValue = { 
    locationObserver,
    history, 
    filterRegistry: filterRegistry.current 
  };

  return (
    <filtersContext.Provider value={contextValue}>
      {children}
    </filtersContext.Provider>
  );
}


export default FiltersProvider;