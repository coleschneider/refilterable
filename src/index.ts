import FiltersProvider from "./FiltersProvider";
import createFilter from "./createFilter";
import useFilter from "./useFilter";
import useReset from "./useReset";

export { useReset, useFilter, createFilter, FiltersProvider };
