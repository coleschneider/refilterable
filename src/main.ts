import * as React from 'react';
import FiltersProvider from "./FiltersProvider";
import createFilter from "./createFilter";
import useFilter from "./useFilter";
import useReset from "./useReset";

export default {
	FiltersProvider,
	createFilter,
	useFilter,
	useReset
};