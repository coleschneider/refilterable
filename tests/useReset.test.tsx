import React from 'react';
import { History } from 'history';
import { renderHook, act } from 'react-hooks-testing-library'
import { useFilter, useReset, FiltersProvider, createFilter } from '../src';
import { createMemoryHistory } from 'history';

describe('useReset', () => {
  let wrapper, history: History;

  beforeEach(() => {
    history = createMemoryHistory();
    
    wrapper = ({ children }) => (
      <FiltersProvider 
        history={history}
      >
        {children}
      </FiltersProvider>
    );
  });

  it('should throw an error when there is no provider in the tree', () => {
    expect(() => {
      renderHook(() => useReset()).result.current;
    }).toThrowErrorMatchingSnapshot();
  });


  it('should delete all the registered filters without default values', () => {
    const { result: hooks } = renderHook(() => [
      useFilter('foo'),
      useFilter('bar'),
      useReset(),
    ], { wrapper });

    act(() => {
      const [fooHook, barHook, reset] = hooks.current;

      // @ts-ignore
      const [, setFoo] = fooHook;
      // @ts-ignore
      const [, setBar] = barHook;

      // @ts-ignore
      setFoo('foo');
      // @ts-ignore
      setBar('bar');

      expect(history.location.search).toBe('?foo=foo&bar=bar');

      // @ts-ignore
      reset();
      expect(history.location.search).toBe('');
      expect(fooHook[0]).toBeUndefined();
      expect(barHook[0]).toBeUndefined();
    });
  });

  it('should reset all the registered filters with default values to their default values', () => {
    const { result: hooks } = renderHook(() => [
      useFilter(createFilter('foo', { defaultValue: 'foo' })),
      useReset(),
    ], { wrapper });

    act(() => {
      const [fooHook, reset] = hooks.current;

      // @ts-ignore
      const [, setFoo] = fooHook;

      // @ts-ignore
      setFoo('bar');

      expect(history.location.search).toBe('?foo=bar');

      // @ts-ignore
      reset();
      expect(history.location.search).toBe('?foo=foo');
      expect(fooHook[0]).toBe('foo');
    });
  });

  it('should not impact any URL parameters managed outside of re-filter', () => {
    history.push({ search: 'legacyParam=10' });
    const { result } = renderHook(() => useReset(), { wrapper });

    act(() => {
      const reset = result.current;
      reset();

    });

    expect(history.location.search).toBe('?legacyParam=10');
  });

  it('should not change the URL if the "dry" option is set to true', () => {
    history.push({ search: 'foo=bar' });
    
    const { result: hooks } = renderHook(() => [
      useFilter('foo'),
      useReset(),
    ], { wrapper });

    let nextSearch;

    act(() => {
      const reset = hooks.current[1];
      // @ts-ignore
      nextSearch = reset({ dry: true });
    });

    // @ts-ignore
    const [foo] = hooks.current[0];

    expect(history.location.search).toBe('?foo=bar');
    expect(foo).toBe('bar');
    expect(nextSearch).toBe('');
  });
  
});