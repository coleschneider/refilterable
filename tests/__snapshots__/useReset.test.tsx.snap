// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`useReset should throw an error if the value provided is not a filter 1`] = `
"refilterable: you called useReset() and passed an invalid filter object.
      Instead of constructing the configuration object on your own,
      use createFilter() or composeFilters()."
`;

exports[`useReset should throw an error when there is no provider in the tree 1`] = `"refilterable: Components that utilize the \\"useFilter\\" hook need to be wrapped with FiltersProvider."`;
